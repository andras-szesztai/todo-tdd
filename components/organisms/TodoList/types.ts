import { TTodoItem } from 'components/molecules/TodoItem'

export interface Props {
    todoItems: TTodoItem[]
}
