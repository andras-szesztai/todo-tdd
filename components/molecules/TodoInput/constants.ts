export const TODO_INPUT_LABEL = 'Todo Title'
export const TODO_INPUT_PLACEHOLDER = 'Enter a Todo title...'
export const TODO_INPUT_BUTTON_TEXT = 'Submit'

export const TODO_INPUT_ID = 'todo-input'
