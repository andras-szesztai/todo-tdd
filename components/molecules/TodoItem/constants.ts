export const TODO_COMPLETE_BUTTON_TEXT = {
    complete: 'Complete',
    unComplete: 'Un-complete',
}
