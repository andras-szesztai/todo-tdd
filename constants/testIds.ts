export const TEST_ID = {
    todoViewContainer: 'todos-view-container',
}
