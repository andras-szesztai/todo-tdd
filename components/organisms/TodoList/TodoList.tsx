import { TodoItem } from 'components/molecules/TodoItem'

import { Props } from './types'

function TodoList({ todoItems, onTodoItemCompleteClick }: Props) {
    const sortedTodoItems = [...todoItems].sort((a, b) => {
        return new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
    })
    return (
        <ul className="flex flex-col gap-2 p-2 pr-4 overflow-y-auto border h-96 w-80 border-primary-dark rounded-xl bg-primary-dark-disabled">
            {sortedTodoItems.map((todoItem) => (
                <TodoItem
                    key={todoItem.id}
                    title={todoItem.title}
                    isCompleted={todoItem.isCompleted}
                    onCompleteClick={() => {
                        onTodoItemCompleteClick(todoItem.id)
                    }}
                />
            ))}
        </ul>
    )
}

export default TodoList
