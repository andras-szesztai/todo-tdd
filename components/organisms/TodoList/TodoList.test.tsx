import { fireEvent, render, within } from '@testing-library/react'
import { TEST_ID } from 'constants/testIds'

import TodoList from './TodoList'

describe('TodoList', () => {
    it('renders without error', () => {
        const { getByTestId, asFragment } = render(
            <TodoList todoItems={[]} onTodoItemCompleteClick={jest.fn} />
        )
        const todoListContainer = getByTestId(TEST_ID.todoListContainer)
        expect(todoListContainer).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('displays submitted todos in the ui as list items ordered by date created ascending', () => {
        const testTodoItems = [
            {
                id: '2',
                title: 'My Todo 2',
                createdAt: '2022-11-28T14:54:38.113Z',
                isCompleted: false,
            },
            {
                id: '1',
                title: 'My Todo 1',
                createdAt: '2022-11-27T14:54:38.113Z',
                isCompleted: false,
            },
        ]
        const { getAllByTestId, asFragment } = render(
            <TodoList
                todoItems={testTodoItems}
                onTodoItemCompleteClick={jest.fn}
            />
        )
        const todoItems = getAllByTestId(TEST_ID.todoItem)
        expect(todoItems).toHaveLength(testTodoItems.length)
        expect(todoItems[0]).toHaveTextContent(testTodoItems[1].title)
        expect(todoItems[1]).toHaveTextContent(testTodoItems[0].title)
        expect(asFragment()).toMatchSnapshot()
    })

    it('calls onTodoItemCompleteClick when a todo item is clicked', () => {
        const testTodoItems = [
            {
                id: '1',
                title: 'My Todo 1',
                createdAt: '2022-11-27T14:54:38.113Z',
                isCompleted: false,
            },
            {
                id: '2',
                title: 'My Todo 2',
                createdAt: '2022-11-28T14:54:38.113Z',
                isCompleted: false,
            },
        ]
        const mockOnTodoItemCompleteClick = jest.fn()
        const { getAllByTestId } = render(
            <TodoList
                todoItems={testTodoItems}
                onTodoItemCompleteClick={mockOnTodoItemCompleteClick}
            />
        )
        const todoItem = getAllByTestId(TEST_ID.todoItem)
        const selectedIndex = 0
        expect(mockOnTodoItemCompleteClick).not.toHaveBeenCalled()
        fireEvent.click(
            within(todoItem[selectedIndex]).getByTestId(
                TEST_ID.todoCompleteButton
            )
        )
        expect(mockOnTodoItemCompleteClick).toHaveBeenCalledTimes(1)
        expect(mockOnTodoItemCompleteClick).toHaveBeenCalledWith(
            testTodoItems[selectedIndex].id
        )
    })
})
