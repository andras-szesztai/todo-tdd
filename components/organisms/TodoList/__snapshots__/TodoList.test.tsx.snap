// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList displays submitted todos in the ui as list items ordered by date created ascending 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col gap-2 p-2 pr-4 overflow-y-auto border h-96 w-80 border-primary-dark rounded-xl bg-primary-dark-disabled"
  >
    <li
      class="relative border border-primary-dark rounded-lg p-2 truncate group bg-white border-primary-dark text-primary-dark"
    >
      My Todo 1
      <span
        class="absolute transition-opacity -translate-y-1/2 opacity-0 right-2 top-1/2 group-hover:opacity-100"
      >
        <button
          class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-xs py-1 px-1.5 rounded-md"
          type="button"
        >
          Complete
        </button>
      </span>
    </li>
    <li
      class="relative border border-primary-dark rounded-lg p-2 truncate group bg-white border-primary-dark text-primary-dark"
    >
      My Todo 2
      <span
        class="absolute transition-opacity -translate-y-1/2 opacity-0 right-2 top-1/2 group-hover:opacity-100"
      >
        <button
          class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-xs py-1 px-1.5 rounded-md"
          type="button"
        >
          Complete
        </button>
      </span>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`TodoList renders without error 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col gap-2 p-2 pr-4 overflow-y-auto border h-96 w-80 border-primary-dark rounded-xl bg-primary-dark-disabled"
  />
</DocumentFragment>
`;
